dataverse/orgsvc/C#-NETCore/BulkOperations/UpsertMultiple/Program.cs

```diff
@@ -0,0 +1,269 @@
+            string tableLogicalName = tableSchemaName.ToLower(); //sample_example
+
+            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
+
+            // Create a Dataverse service client using the default connection string.
+            ServiceClient serviceClient =
+                new(app.Configuration.GetConnectionString("default"))
+                {
+                    UseWebApi = false
+                };
+
+            try
+            {
+                // Create sample_Example table for this sample
+                Utility.CreateExampleTable(
+                    serviceClient: serviceClient,
+                    tableSchemaName: tableSchemaName,
+                    isElastic: Settings.UseElastic);
+
+                // Confirm the table supports UpsertMultiple
+                if (!Utility.IsMessageAvailable(
+                    service: serviceClient,
+                    entityLogicalName: tableLogicalName,
+                    messageName: "UpsertMultiple"))
+                {
+                    Console.WriteLine($"The UpsertMultiple message is not available " +
+                        $"for the {tableSchemaName} table.");
+                }
+                else
+                {
+                    // For UpsertMultiple, we will try to set the batch to contain records
+                    // where half of them will be created and half will be updated by UpsertMultiple
+                    int nCreate = numberOfRecords / 2;
+                    EntityCollection entities = new();
+
+                    var createdRecordIds = CreateRecordsUtility(nCreate, tableLogicalName, serviceClient, out entities);
+
+                    Console.WriteLine($"\nPreparing {nCreate} records to update..");
+
+                    // Assign Id values to created items to prepare to update them
+                    for (int i = 0; i < createdRecordIds.Length; i++)
+                    {
+                        entities.Entities[i].Id = createdRecordIds[i];
+                    }
+
+                    // Update the sample_name value:
+                    foreach (Entity entity in entities.Entities)
+                    {
+                        entity["sample_name"] += " Updated";
+                    }
+
+                    Console.WriteLine($"\nPreparing {nCreate} records to create..");
+                    for (int i = 0; i < nCreate; i++)
+                    {
+                        entities.Entities.Add(new Entity(tableLogicalName)
+                        {
+                            Attributes = {
+                            // Example: 'sample record 0000001'
+                            { "sample_name", $"sample record {i+1:0000000}" }
+                        }
+                        });
+                    }
+
+                    #region Send UpsertMultipleRequest
+                    // Use UpsertMultipleRequest
+                    UpsertMultipleRequest upsertMultipleRequest = new()
```

Comment:
@SomDaDe 

Upsert and alternate keys need to be demonstrated together.

When creating the table, please add another string column, perhaps called `sample_externalid`, which will contain some unique value.

Use [CreateEntityKeyRequest](https://learn.microsoft.com/en-us/dotnet/api/microsoft.xrm.sdk.messages.createentitykeyrequest?view=dataverse-sdk-latest) to create the key.
The value could be `Guid.NewGuid().ToString("n")`, a GUID without hyphens.  Don't need to get fancy.

Then, rather than assign the primary key GUID values, assign the alternate key values and use them in UpsertMultiple.

In the `CreateRecordsUtility`, return the records with the alternate key value ONLY..

Reply:
Done